docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,292 @@
+itself and cluster wide strategies enforced by an allocator.  The separation of concerns falls into
+two distinct categories: permissions that the pod can request and runtime parameters the cluster
+admin controls.
+
+An example of a permission that can be requested is to run a pod as privileged.  This is set on a
+pod by pod basis.  A cluster wide setting may be the allocation of user ids that pods must run under.
+The cluster administrator may decide that no pod can run as the root user in their cluster and
+provide a block of UIDs that the allocator can use to distribute amongst namespaces.
+
+A security context constraint object exists in the root scope, outside of a namespace.  The
+security context constraints can reference users, groups, or a service account that is allowed
+to operate under the constraints.
+
+A one to one relationship will be maintained for a security context constraint and a service account.
+Users may be allowed to have many security context constraints and admission will create a unioned
+set of their highest allowed permissions when creating a pod (discussed in admission, below).
+
+To minimize the amount of objects that must be managed by a cluster administrator the following
+security context constraints objects can be provided as a default set:
+
+1.  User/Group Constraints - this is the default security context constraints for the cluster and is usually the
+most restrictive set of constraints.
+1.  Service Account Constraints - security context constraints for service accounts will be created that allows
+pods to run with different permissions than the user.  The user should be required to have
+permissions to reference the service account in order to run a pod with the service account's
+security context constraints.
+1.  Administrator Constraints - a set of constraints with elevated privileges that can be used
+by an administrative user or group.
+1.  Infrastructure Component Constraints - constraints that may have elevated privileges within the
+cluster but possibly not as elevated as administrator constraints.
+
+```go
+// SecurityContextConstraints governs the ability to make requests that affect the SecurityContext that will
+// be applied to a container.
+type SecurityContextConstraints struct {
+  // Type distinguishes the type of SecurityContextConstraints object.  Acceptable values are
+  // INFRA, USER, DEFAULT_USER, SERVICE_ACCOUNT, DEFAULT_SERVICE_ACCOUNT
+  Type string
+  // AllowPrivilegedContainer determines if a container can request to be run as privileged.
+  AllowPrivilegedContainer bool
+  // AllowedCapabilities is a list of capabilities that can be requested to add to the container.
+  AllowedCapabilities []CapabilityType
+  // AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
+  AllowHostDirVolumePlugin bool
+  // The users who have permissions to use this security context constraints
+  AllowedUsers []string
```

Comment:
I really don't want to distinguish between users and service accounts here... authentication and authorization layers don't. Rather than `AllowedUsers`/`AllowedServiceAccounts`, I'd rather see:
```
Users []string
Groups []string
```

That lets you treat service accounts and users uniformly to determine the applicable SecurityContextConstraints, as long as you can map a service account to a username/group list. That mapping already has to be done for the authentication/authorization layers.

Group support could make it easy to express all of the cases described as we build out group support:
* The default constraints object could include a `system:authenticated` group imposed by the authentication layer on all authenticated users
* Default constraints for service accounts could include a `system:service-accounts` group imposed by the service account authenticator / token generator.
* Cluster admin constraints could include a `system:cluster-admins` group
* Infrastructure components running in a privileged namespace could be identified by a particular service account (e.g. `serviceaccount:infra:builder`), or a particular group containing all service accounts for a namespace (e.g. `serviceaccounts:infra`)